src/shared/lib/GenericAssayUtils/GenericAssayCommonUtils.ts

```diff
@@ -1 +1,132 @@
+    molecularProfiles: MolecularProfile[]
+) {
+    const genericAssayProfiles = molecularProfiles.filter(profile => {
+        return (
+            profile.molecularAlterationType ===
+            AlterationTypeConstants.GENERIC_ASSAY
+        );
+    });
+
+    const genericAssayProfilesGroupByGenericAssayType = _.groupBy(
+        genericAssayProfiles,
+        'genericAssayType'
+    );
+    const genericAssayTypes = _.keys(
+        genericAssayProfilesGroupByGenericAssayType
+    );
+
+    const GenericAssayMetaGroupByGenericAssayType: {
+        [genericAssayType: string]: GenericAssayMeta[];
+    } = {};
+
+    await Promise.all(
+        genericAssayTypes.map(async genericAssayType => {
+            const genericAssayProfiles =
+                genericAssayProfilesGroupByGenericAssayType[genericAssayType];
+            const genericAssayProfileIds = genericAssayProfiles.map(
+                profile => profile.molecularProfileId
+            );
+            GenericAssayMetaGroupByGenericAssayType[
+                genericAssayType
+            ] = await fetchGenericAssayMetaByProfileIds(genericAssayProfileIds);
+        })
+    );
+    return GenericAssayMetaGroupByGenericAssayType;
+}
+
+export async function fetchGenericAssayMetaByMolecularProfileIdsGroupByMolecularProfileId(
+    molecularProfiles: MolecularProfile[]
+) {
+    const genericAssayProfiles = molecularProfiles.filter(profile => {
+        return (
+            profile.molecularAlterationType ===
+            AlterationTypeConstants.GENERIC_ASSAY
+        );
+    });
+
+    const genericAssayMetaGroupByMolecularProfileId: {
+        [molecularProfileId: string]: GenericAssayMeta[];
+    } = {};
+
+    await Promise.all(
+        genericAssayProfiles.map(async genericAssayProfile => {
+            const genericAssayProfileId =
+                genericAssayProfile.molecularProfileId;
+            genericAssayMetaGroupByMolecularProfileId[
+                genericAssayProfileId
+            ] = await fetchGenericAssayMetaByProfileIds([
+                genericAssayProfileId,
+            ]);
+        })
+    );
+    return genericAssayMetaGroupByMolecularProfileId;
+}
+
+export async function fetchGenericAssayMetaByProfileIds(
+    genericAssayProfileIds: string[]
+) {
+    let genericAssayMeta: GenericAssayMeta[] = [];
+    if (genericAssayProfileIds.length > 0) {
+        genericAssayMeta = await client.fetchGenericAssayMetaDataUsingPOST({
+            genericAssayMetaFilter: {
+                molecularProfileIds: genericAssayProfileIds,
+                // the Swagger-generated type expected by the client method below
+                // incorrectly requires both molecularProfileIds and genericAssayStableIds;
+                // use 'as' to tell TypeScript that this object really does fit.
+            } as GenericAssayMetaFilter,
+        });
+    }
+    return genericAssayMeta;
+}
+
+export async function fetchGenericAssayMetaByEntityIds(entityIds: string[]) {
```

Comment:
here, there's no reason to use async/await since it doesn't improve the clarity of your code.  just the the result of client.fetchGenericAssayMetaDataUsingPost which is nicely typed promise.    or, if entetyIds.length = 0, Promise.resolve([])

Reply:
same as above.  i think it's a good idea not to use async on fetch functions but instead always return promise.

Reply:
agree here but not above because in the case up there it was about populating map entries which I don't see as clean a way to do in the way you're suggesting